Src/WindowsServer/WindowsServer.Shared/Implementation/AzureMetadataRequestor.cs

```diff
@@ -0,0 +1,113 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Net;
+    using System.Net.Http;
+    using System.Runtime.Serialization.Json;
+    using System.Threading.Tasks;
+#if NETSTANDARD1_3
+    using System.Net.Http;
+#endif
+    using Extensibility;
+    using Microsoft.ApplicationInsights.WindowsServer.Implementation.DataContracts;
+
+    internal class AzureMetadataRequestor : IAzureMetadataRequestor
+    {
+        /// <summary>
+        /// Azure Instance Metadata Service exists on a single non-routable IP on machines configured
+        /// by the Azure Resource Manager. See <a href="https://go.microsoft.com/fwlink/?linkid=864683">to learn more.</a>
+        /// </summary>
+        internal static string AzureImsApiVersion = $"api-version=2017-08-01"; // this version has the format=text capability
+        internal static string AzureImsJsonFormat = "format=json";
+        internal static int AzureImsMaxResponseBufferSize = 512;
+
+        /// <summary>
+        /// Default timeout for the web requests made to obtain Azure IMS data. Internal to expose to tests.
+        /// </summary>
+        internal TimeSpan AzureImsRequestTimeout = TimeSpan.FromSeconds(10);
+
+        /// <summary>
+        /// Private function for mocking out the actual call to IMS in unit tests. Available to internal only.
+        /// </summary>
+        private Func<string, Task<AzureInstanceComputeMetadata>> azureIMSRequestor = null;
+
+        private string baseImdsUri = $"http://169.254.169.254/metadata/instance/compute";
+
+        internal AzureMetadataRequestor(Func<string, Task<AzureInstanceComputeMetadata>> makeAzureIMSRequestor = null)
+        {
+            this.azureIMSRequestor = makeAzureIMSRequestor;
+        }
+
+        /// <summary>
+        /// Gets or sets the base URI for the Azure Instance Metadata service. Internal to allow overriding in test.
+        /// </summary>
+        internal string BaseAimsUri
+        {
+            get => this.baseImdsUri;
+            set => this.baseImdsUri = value;
+        }
+
+        public async Task<AzureInstanceComputeMetadata> GetAzureComputeMetadata()
+        {
+            string metadataRequestUrl = $"{this.BaseAimsUri}?{AzureMetadataRequestor.AzureImsJsonFormat}&{AzureMetadataRequestor.AzureImsApiVersion}";
+
+            AzureInstanceComputeMetadata jsonResponse = await this.MakeAzureMetadataRequest(metadataRequestUrl);
+
+            return jsonResponse;
+        }
+
+        private async Task<AzureInstanceComputeMetadata> MakeAzureMetadataRequest(string metadataRequestUrl)
+        {
+            AzureInstanceComputeMetadata requestResult = null;
+
+            SdkInternalOperationsMonitor.Enter();
+            try
+            {
+                if (this.azureIMSRequestor != null)
+                {
+                    requestResult = await this.azureIMSRequestor(metadataRequestUrl);
```

Comment:
can we simplify the code here by just calling the this.azureIMSRequestor() - and let constructor worry about which method this.azureIMSRequestor points to?

Reply:
This is an interesting piece of code – I went with what I saw as readability over simplicity in code-calls. I wanted to be very clear to the reader of this code what was going on. I can be convinced to simplify it here  as you suggest but I’d like to discuss further first. Please raise an issue to be addressed in beta2 if you wish.